lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb

```diff
@@ -0,0 +1,162 @@
+      class EmulatorHelper
+        BOOT_WAIT = 2
+        BOOT_TIMEOUT = 60
+
+        SHUTDOWN_WAIT = 2
+        SHUTDOWN_TIMEOUT = 60
+
+        def initialize
+          @tools = Fastlane::Helper::Android::ToolsPathHelper.new
+        end
+
+        # Installs the system-image suitable for a given Android `api`, with `google_apis`, and for the current machine's architecture
+        #
+        # @param [Integer] api The Android API level to use
+        #
+        # @return [String] The `sdkmanager` package specifier that has been installed
+        #
+        def install_system_image(api:)
+          package = system_image_package(api: api)
+
+          UI.message("Installing System Image for Android #{api} (#{package})")
+          Actions.sh(@tools.sdkmanager, '--install', package)
+          UI.success("System Image #{package} successfully installed.")
+          package
+        end
+
+        # Create an emulator (AVD) for a given `api` number and `device` model
+        #
+        # @param [Integer] api The Android API level to use for this AVD
+        # @param [String] device The Device Model to use for this AVD. Valid values can be found using `avdmanager list devices`
+        # @param [String] name The name to give for the created AVD. Defaults to `<device>_API_<api>`.
+        # @param [String] sdcard The size of the SD card for this device. Defaults to `512M`.
+        #
+        # @return [String] The device name (i.e. either `name` if provided, or the derived `<device>_API_<api>` if provided `name` was `nil``)
+        #
+        def create_avd(api:, device:, system_image: nil, name: nil, sdcard: '512M')
+          package = system_image || system_image_package(api: api)
+          device_name = name || "#{device.gsub(' ', '_').capitalize}_API_#{api}"
+
+          UI.message("Creating AVD `#{device_name}` (#{device}, API #{api})")
+
+          Actions.sh(
+            @tools.avdmanager, 'create', 'avd',
+            '--force',
+            '--package', package,
+            '--device', device,
+            '--sdcard', sdcard,
+            '--name', device_name
+          )
+
+          UI.success("AVD `#{device_name}` successfully created.")
+
+          device_name
+        end
+
+        # Launch the emulator for the given AVD, then return the emulator serial
+        #
+        # @param [String] name name of the AVD to launch
+        # @param [Boolean] cold_boot if true, will do a cold boot, if false will try to use a previous snapshot of the device
+        # @param [Boolean] wipe_data if true, will wipe the emulator (i.e. reset the user data image)
+        #
+        # @return [String] emulator serial number corresponding to the launched AVD
+        #
+        def launch_avd(name:, cold_boot: true, wipe_data: true)
+          port = '5554'.freeze
+          serial = "emulator-#{port}"
+
+          shut_down_emulators!(serials: [serial]) # To ensure we can launch one on the port 5554 (as it's hardcoded for simplicity)
```

Comment:
Known limitation of the current implementation: the port is hardcoded to `5554` — which is why we need to ensure we shut down existing emulators if any first — because it ended up not being that trivial to make it dynamic.

If I let `emulator` choose the ports by itself (instead of explicitly providing them via `-port` flag on line 77), which would ideally be a more flexible alternative, I then couldn't find an easy and consistent/safe enough way to then retrieve the port that was used. I considered looking at the device list before and after to compare and detect the new one, but:
 - There's a small risk that the new emulator disappears from `adb devices -l` for a brief moment while it's booting and setting itself up during booting sequence
 -  That's not a strong guarantee because the user might also have other emulators that are in the process of booting or shutting down in parallel, which would lead to some race condition on checking the device list to compare
 - I've also took a quick look at if there were some command that would allow me to get a list of AVD names <-> adb ports used, which maybe we could use to try and match the AVD name with its port.
    - I've found some ideas around looping on each emulator, and call `adb -s emulator-#{port} emu avd name` command on each to get the name of the AVD…
    - …but that supposes the emulator has already finished booting (and we can't use `adb -s emulator-#{port} wait-for-device` to wait for the AVD we just launched to finish being ready… because we need to know the serial and thus the port to call that one… so catch-22)
    - …and we could consider looping over all the emulators found in `adb devices -l`, and call `adb -s emulator-#{port}` on each… but if we happen to do that loop before the AVD we just launched hasn't finished booting and connecting to `adb`, that won't find it and we'd have to do the loop all over again (using our `retry_loop` helper, probably), and calling `adb emu avd name` all over again… doesn't feel very efficient nor stable?

So in the end, I'm open to any ideas on how to improve flexibility here, or ways to find out the serial of an AVD we just booted without risking falling into the cracks and race condition / timing traps highlighted above. (@ParaskP7  any ideas or `adb` magic commands you know could help?)

---

The impact of this known limitation of hard-coding the port for now is that we won't be able to run screenshot tests on the same machine in parallel (as those would try to boot multiple emulators). If we find a working solution to make it possible for it to be dynamic (and thus allow screenshot parallelism), we can always add the feature in a subsequent PR.

Reply:
EDIT: I've ended up fixing that one with latest commits on this PR after all 🎉